packages/build-info/src/workspaces.ts

```diff
@@ -1,25 +1,85 @@
+import { existsSync, readFileSync } from 'fs'
+import { join, relative } from 'path'
+
 import mapWorkspaces from '@npmcli/map-workspaces'
+import { PackageJson } from 'read-pkg'
+import { parse } from 'yaml'
 
 import type { Context } from './context.js'
+import { PkgManager, PkgManagerFields } from './detect-package-manager.js'
+
+export type WorkspaceInfo = {
+  /** if we are in the current workspace root or not */
+  isRoot: boolean
+  /** the workspace root directory */
+  rootDir: string
+  /** list of relative package paths inside the workspace */
+  packages: string[]
+}
+
+/**
+ * Get a list of globs about all the packages inside a pnpm workspace
+ * https://pnpm.io/pnpm-workspace_yaml
+ */
+export const detectPnpmWorkspaceGlobs = (rootDir: string): string[] => {
+  const workspaceFile = join(rootDir, 'pnpm-workspace.yaml')
+  if (!existsSync(workspaceFile)) {
+    return []
+  }
+
+  const { packages } = parse(readFileSync(workspaceFile, 'utf-8'))
+  return packages
+}
 
-export const getWorkspaceInfo = async (context: Context) => {
-  if (!context.rootPackageJson.workspaces) {
+export const detectNpmOrYarnWorkspaceGlobs = (pkgJson: PackageJson): string[] => {
+  if (Array.isArray(pkgJson.workspaces)) {
+    return pkgJson.workspaces
+  }
+  if (typeof pkgJson.workspaces === 'object') {
```

Comment:
The documentation implies that the workspace field is an array. When could it be an object?

> The optional workspaces field is an array of file patterns

Reply:
For NPM yes but not for yarn:
![CleanShot 2022-10-25 at 12 06 44](https://user-images.githubusercontent.com/11156362/197745750-471d6f5e-1a28-4675-a77c-378d2801cc53.png)

![CleanShot 2022-10-25 at 12 07 45](https://user-images.githubusercontent.com/11156362/197745939-6a48f61d-0301-4427-974e-136f77a3ff30.png)